src/mlpack/methods/ann/quantize.hpp

```diff
@@ -0,0 +1,60 @@
+#ifndef MLPACK_METHODS_ANN_QUANTIZATION_QUANTIZE_HPP
+#define MLPACK_METHODS_ANN_QUANTIZATION_QUANTIZE_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/methods/ann/ffn/ffn.hpp>
+#include <mlpack/methods/ann/layer/layer.hpp>
+#include "quantization_utils.hpp"
+
+namespace mlpack {
+namespace ann {
+
+/**
+ * Quantize a neural network from one matrix type to another.
+ *
+ * @tparam TargetType The target element type for quantization (e.g., int8_t).
+ * @tparam NetworkType The type of the original network.
+ * @param network The original network to be quantized.
+ * @return A new network with weights quantized to the TargetType.
+ */
+template<typename TargetType, typename NetworkType>
+auto Quantize(const NetworkType& network)
```

Comment:
I guess that this is the intended top-level function for the user to use?

When I read through the member function `FFN::Quantize()`, I thought that that could be another way to go for a user-facing API.  So it seems to me like either of these are good enough choices for a user:

 * `FFN<..., arma::imat> = network.Quantize<arma::imat>();`
 * `FFN<..., arma::imat> = Quantize<arma::mat>(network);`

It seems equivalent enough.  However I suspect that the quantization itself will need to take place in a separate utilities directory or file, so that the same code could also be applied to a function `RNN::Quantize()` (or similar).

As far as design goes, I only have one other major concern, which is: I think that there are some quantization techniques out there that will quantize each layer differently.  Do you have any ideas how we might support that from both a user-facing API perspective and also an internal implementation perspective?  Ideally such a design should not involve putting quantization-related code in each layer's implementation.  (Maybe it is worth finding some papers that do layer-specific quantization just to get an idea of what the range of strategies we might want to someday support is.)